### PR TITLE
Implement `mode` option for `dotnet publish`.

### LINKED ISSUE #6237: dotnet new console, publish, makes a dll not an exe
```
C:\PLAy\asdf>dotnet --version
2.0.0-preview1-005694

C:\PLAy\asdf>dotnet new console
The template "Console Application" was created successfully.

C:\PLAy\asdf>dotnet restore

...
C:\PLAy\asdf>dotnet run
Hello World!

C:\PLAy\asdf>dotnet publish
Microsoft (R) Build Engine version 15.2.47.30403
Copyright (C) Microsoft Corporation. All rights reserved.

  asdf -> C:\PLAy\asdf\bin\Debug\netcoreapp2.0\asdf.dll
```

Shouldn't I have a console .exe? How do I run this dll?

### PR DESCRIPTION
**Pull request by peterhuene**
This PR implements a `mode` option that can control how an application is
published with the `dotnet publish` command.

There are three supported modes:

* `self-contained`: publishes a self-contained application (same as
--self-contained).
* `fx-dependent`: publishes a framework-dependent application (with an
application host when a runtime is specified).
* `fx-dependent-no-exe`: publishes a framework-dependent application without an
application host.

The default when publishing without a runtime specified is
`fx-dependent-no-exe`.

The default when publishing with a runtime specified is `self-contained`.

`fx-dependent` requires netcoreapp2.1 or later when a runtime is specified.

The `--self-contained` option is still supported, but is now hidden so that
users will be encouraged to move to the `--mode` option.

Fixes #6237.

### CODE REVIEW

File: test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs [hunk 1]
```diff
@@ -165,36 +183,25 @@ public void ItPublishesARidSpecificAppSettingSelfContainedToFalse()
                      .And.HaveStdOutContaining("Hello World");
         }
 
-        private DirectoryInfo PublishAppWithSelfContained(string testAppName, bool selfContained)
+        private DirectoryInfo PublishApp(string testAppName, string rid, string mode, string args = null)
         {
             var testInstance = TestAssets.Get(testAppName)
-                .CreateInstance($"PublishesSelfContained{selfContained}")
+                .CreateInstance($"PublishApp_{rid ?? "none"}_{mode ?? "none"}")
```
Comment (dsplaisted) [inline test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs]: Does `args` need to be included here as well?

File: src/dotnet/commands/dotnet-publish/LocalizableStrings.resx [hunk 2]
```diff
@@ -135,9 +135,16 @@
   <data name="ManifestOptionDescription" xml:space="preserve">
     <value>The path to a target manifest file that contains the list of packages to be excluded from the publish step.</value>
   </data>
-  <data name="SelfContainedOptionDescription" xml:space="preserve">
-    <value>Publish the .NET Core runtime with your application so the runtime doesn't need to be installed on the target machine.
-The default is 'true' if a runtime identifier is specified.</value>
+  <data name="ModeOptionDescription" xml:space="preserve">
+    <value>The mode to use when publishing the application.
+The 'self-contained' mode publishes the application with the .NET Core runtime.
+The 'fx-dependent' mode publishes the application as framework-dependent with an executable if a target runtime is specified.
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]: Pushed.

File: src/dotnet/commands/dotnet-publish/LocalizableStrings.resx [hunk 3]
```diff
@@ -146,10 +153,16 @@ The default is 'true' if a runtime identifier is specified.</value>
...
-    <value>The target runtime to publish for. This is used when creating self-contained deployment.
-The default is to publish a framework-dependent application.</value>
+    <value>The target runtime to publish the application for.
+The default is to publish a framework-dependent application without an executable.</value>
   </data>
   <data name="ConfigurationOptionDescription" xml:space="preserve">
     <value>The configuration to publish for. The default for most projects is 'Debug'.</value>
   </data>
+  <data name="PublishModeAndSelfContainedOptionsConflict" xml:space="preserve">
+    <value>The '--mode' and '--self-contained' options conflict. Specify only one of the options.</value>
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply src/dotnet/commands/dotnet-publish/LocalizableStrings.resx]: Pushed.

File: test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs (hunk 1 above)
Comment (peterhuene) [reply test/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs]: Pushed.

**Comment by dasMulli** [comment]
About the CLI logic vs MSBuild logic:
Everything could be done in both places. It gets more interesting when thinking about scenarios that could include adding additional targets, e.g. `/t:LinkNative`.
The mode is more or less "syntactic sugar" to easily use the available options. The only benefit of doing it in MSBuild is that it is easier to use for direct `msbuild` invocation. Currently, one needs to look up the parser code to find out which CLI options correspond to which MSBuild property - or find it in the docs or StackOverflow (`PackageOutputPath` for example).

**Comment by wli3** [comment]
@dsplaisted thanks! looks this is all green, should we merge it? @livarcocc

**Comment by livarcocc** [comment]
Talked to @peterhuene and we will wait for @nguerrera to be back and take a look before merging.

**Comment by wli3** [comment]
Talked to @livarcocc later. This is blocking info bar change. So, let's merge it first and follow up later.

**Comment by wli3** [comment]
Added https://github.com/dotnet/cli/issues/9503 to keep track

**Pull request merged**